src/main/java/com/onarandombox/MultiverseCore/utils/player/PlayerActionChecker.java

```diff
@@ -0,0 +1,242 @@
+package com.onarandombox.MultiverseCore.utils.player;
+
+import com.onarandombox.MultiverseCore.MultiverseCore;
+import com.onarandombox.MultiverseCore.api.MVWorld;
+import com.onarandombox.MultiverseCore.api.MVWorldManager;
+import com.onarandombox.MultiverseCore.api.action.ActionResponse;
+import com.onarandombox.MultiverseCore.destination.ParsedDestination;
+import com.onarandombox.MultiverseCore.utils.PermissionsTool;
+import com.onarandombox.MultiverseCore.utils.player.checkresult.BlacklistResult;
+import com.onarandombox.MultiverseCore.utils.player.checkresult.EntryFeeResult;
+import com.onarandombox.MultiverseCore.utils.player.checkresult.GameModeResult;
+import com.onarandombox.MultiverseCore.utils.player.checkresult.NullPlaceResult;
+import com.onarandombox.MultiverseCore.utils.player.checkresult.PlayerLimitResult;
+import com.onarandombox.MultiverseCore.utils.player.checkresult.UseDestinationResult;
+import com.onarandombox.MultiverseCore.utils.player.checkresult.WorldAccessResult;
+import com.onarandombox.MultiverseCore.world.configuration.EntryFee;
+import org.bukkit.Location;
+import org.bukkit.Material;
+import org.bukkit.command.BlockCommandSender;
+import org.bukkit.command.CommandSender;
+import org.bukkit.command.ConsoleCommandSender;
+import org.bukkit.entity.Player;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Checks if a player can perform an action.
+ */
+public class PlayerActionChecker {
+    private final MultiverseCore plugin;
+    private final MVWorldManager worldManager;
+    private final PermissionsTool permissionsTool;
+
+    /**
+     * Creates a new PlayerActionChecker.
+     *
+     * @param plugin The MultiverseCore plugin.
+     */
+    public PlayerActionChecker(@NotNull MultiverseCore plugin) {
+        this.plugin = plugin;
+        this.worldManager = plugin.getMVWorldManager();
+        this.permissionsTool = plugin.getPermissionsTool();
+    }
+
+    /**
+     * Checks if a player have permissions to teleport to a destination by another teleporter.
+     *
+     * @param teleporter    The player who is teleporting the teleportee.
+     * @param teleportee    The player to teleport.
+     * @param destination   The destination to teleport to.
+     * @return The result of the check.
+     */
+    public UseDestinationResult canUseDestinationToTeleport(@NotNull CommandSender teleporter,
```

Comment:
Why are some of the return types here `Result` and some `Response`?

Reply:
Response is multiple `ActionResult` object chained, see `ActionResponse` class. Main purpose is so we can get more details of the check and not just a boolean.

Reply:
That relationship is hard to notice from a distance. Could we maybe use a name like `ActionResultChain` or something that makes that notion clearer?

Reply:
Ultimately though, I think the best case would be to just think about `ActionResult` and if it chains, that should hidden in the implementation unless there's a really good reason that other things need to know that the result is a chain or results rather than just if it's successful or not.

Reply:
I think it makes it clearer that this function will return more than 1 result, vs another function that will only return 1 result.

But i do agree with the class naming, made that changed alr.